File: tools/ide/SharpeeIDE/World/WorldMapView.swift
// WorldMapView.swift
// The Map view (ADR-321 D7): the story's rooms drawn on the compass grid the
// analyzer solved, one band per vertical level, with the connections between
// them. The solver's own notes ride along — a room it had to displace to avoid a
// collision, a cycle whose geometry disagrees with itself, a room it could not
// place at all — because those are the cases where what is drawn is not quite
// what the author wrote, and saying so is cheaper than being quietly wrong.
//
// LEVELS ARE BANDS, NOT A PICKER. `up`/`down` exits give a story a z axis, and
// most stories use one or two levels for a handful of rooms; a level control
// would hide most of a small map behind a click. Bands share one x range so the
// levels line up, and a connection between levels shows as a chevron on both
// rooms rather than a line that would have to cross a band boundary.
// Public interface: WorldMapView.show(map:unreached:), WorldMapCanvas.
// Owner context: tools/ide — World.

import AppKit

final class WorldMapView: NSView {

    private let scrollView = NSScrollView()
    private let canvas = WorldMapCanvas()
    private let noteLabel = NSTextField(labelWithString: "")
    private let placeholder = NSTextField(labelWithString: "This story places no rooms yet.")

    override init(frame frameRect: NSRect) {
        super.init(frame: frameRect)

        scrollView.documentView = canvas
        scrollView.hasVerticalScroller = true
        scrollView.hasHorizontalScroller = true
        scrollView.drawsBackground = false
        scrollView.contentView.drawsBackground = false

        noteLabel.font = NSFont.systemFont(ofSize: 10)
        noteLabel.textColor = Theme.foregroundDim
        noteLabel.lineBreakMode = .byWordWrapping
        noteLabel.maximumNumberOfLines = 0
        noteLabel.isHidden = true

        placeholder.font = NSFont.systemFont(ofSize: 11)
        placeholder.textColor = Theme.foregroundFaint
        placeholder.alignment = .center
        placeholder.isHidden = true

        // Wrapping labels must never dictate the pane's width (the divider fight).
        for label in [noteLabel, placeholder] {
            label.setContentCompressionResistancePriority(.defaultLow, for: .horizontal)
        }

        for subview in [scrollView, noteLabel, placeholder] as [NSView] {
            subview.translatesAutoresizingMaskIntoConstraints = false
            addSubview(subview)
        }

        NSLayoutConstraint.activate([
            scrollView.topAnchor.constraint(equalTo: topAnchor),
            scrollView.leadingAnchor.constraint(equalTo: leadingAnchor),
            scrollView.trailingAnchor.constraint(equalTo: trailingAnchor),
            scrollView.bottomAnchor.constraint(equalTo: noteLabel.topAnchor, constant: -4),

            noteLabel.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10),
            noteLabel.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -10),
            noteLabel.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -6),

            placeholder.centerXAnchor.constraint(equalTo: centerXAnchor),
            placeholder.centerYAnchor.constraint(equalTo: centerYAnchor),
        ])
    }

    required init?(coder: NSCoder) {
        fatalError("WorldMapView is not Storyboard-instantiable")
    }

    /// Draws one story's map.
    /// - Parameters:
    ///   - map: the analyzer's solved layout
    ///   - unreached: rooms Reach could not arrive at, drawn faded
    func show(map: WorldMap, unreached: Set<String>) {
        canvas.show(map: map, unreached: unreached)
        let note = Self.solverNote(for: map)
        noteLabel.stringValue = note
        noteLabel.isHidden = note.isEmpty
        let empty = map.positions.isEmpty
        placeholder.isHidden = !empty
        scrollView.isHidden = empty
    }

    /// What the solver had to do to draw this map, as one line, or "" when it
    /// drew the author's geometry exactly.
    ///
    /// A displaced room is never also a skew — resolving a collision moves a room
    /// off its compass cell by construction, which is why the two are counted
    /// separately rather than summed.
    ///
    /// - Parameter map: the analyzer's solved layout
    /// - Returns: the note, or an empty string when there is nothing to say
    static func solverNote(for map: WorldMap) -> String {
        var parts: [String] = []
        if !map.collisions.isEmpty {
            let rooms = map.collisions.map(\.room).joined(separator: ", ")
            parts.append(map.collisions.count == 1
                ? "1 room was moved off its compass cell to avoid an overlap: \(rooms)"
                : "\(map.collisions.count) rooms were moved off their compass cells to avoid overlaps: \(rooms)")
        }
        if !map.skews.isEmpty {
            let rooms = map.skews.map(\.room).joined(separator: ", ")
            parts.append("the geometry disagrees with itself at \(rooms)")
        }
        if !map.unplaced.isEmpty {
            parts.append("not drawn, because no followed exit reaches them: \(map.unplaced.joined(separator: ", "))")
        }
        return parts.joined(separator: " · ")
    }
}

/// The drawing surface itself: rooms, connections, and level bands.
final class WorldMapCanvas: NSView {

    /// One room box's size.
    private static let boxSize = NSSize(width: 108, height: 34)
    /// Gap between adjacent boxes, horizontally and vertically.
    private static let gap = NSSize(width: 24, height: 22)
    /// Margin around the whole drawing.
    private static let margin: CGFloat = 16
    /// Height reserved above each band for its label.
    private static let bandHeader: CGFloat = 20
    /// Gap between two level bands.
    private static let bandGap: CGFloat = 18

    private var map = WorldMap(start: nil, positions: [], unplaced: [],
                               collisions: [], skews: [], connections: [])
    private var unreached: Set<String> = []
    private var displaced: Set<String> = []
    private var frames: [String: NSRect] = [:]
    private var levels: [Int] = []
    private var contentSize = NSSize(width: 1, height: 1)

    /// Rooms are laid out top-down within a band, so the view draws flipped.
    override var isFlipped: Bool { true }

    override var intrinsicContentSize: NSSize { contentSize }

    /// Lays out and redraws one story's map.
    /// - Parameters:
    ///   - map: the analyzer's solved layout
    ///   - unreached: rooms Reach could not arrive at, drawn faded
    func show(map: WorldMap, unreached: Set<String>) {
        self.map = map
        self.unreached = unreached
        self.displaced = Set(map.collisions.map(\.room))
        layoutBoxes()
        invalidateIntrinsicContentSize()
        setFrameSize(contentSize)
        needsDisplay = true
    }

    /// Computes every room's frame and the canvas size.
    ///
    /// Levels stack highest-first; all bands share one x origin so a room above
    /// another sits above it on screen too.
    private func layoutBoxes() {
        frames = [:]
        levels = Array(Set(map.positions.map(\.cell.z))).sorted(by: >)
        guard !map.positions.isEmpty else {
            contentSize = NSSize(width: 1, height: 1)
            return
        }

        let minX = map.positions.map(\.cell.x).min() ?? 0
        let maxX = map.positions.map(\.cell.x).max() ?? 0
        let pitchX = Self.boxSize.width + Self.gap.width
        let pitchY = Self.boxSize.height + Self.gap.height

        var cursorY = Self.margin
        for level in levels {
            let inLevel = map.positions.filter { $0.cell.z == level }
            let maxY = inLevel.map(\.cell.y).max() ?? 0
            let minY = inLevel.map(\.cell.y).min() ?? 0
            cursorY += Self.bandHeader
            for placed in inLevel {
                // North is +y in the analyzer and up on screen: rows count down
                // from the band's northernmost.
                let row = CGFloat(maxY - placed.cell.y)
                let column = CGFloat(placed.cell.x - minX)
                frames[placed.room] = NSRect(
                    x: Self.margin + column * pitchX,
                    y: cursorY + row * pitchY,
                    width: Self.boxSize.width,
                    height: Self.boxSize.height)
            }
            cursorY += CGFloat(maxY - minY + 1) * pitchY + Self.bandGap
        }

        contentSize = NSSize(
            width: Self.margin * 2 + CGFloat(maxX - minX + 1) * pitchX - Self.gap.width,
            height: cursorY - Self.bandGap + Self.margin)
    }

    override func draw(_ dirtyRect: NSRect) {
        Theme.playBackground.setFill()
        dirtyRect.fill()
        drawBandLabels()
        drawConnections()
        drawRooms()
    }

    /// Labels each level band, when the story has more than one.
    private func drawBandLabels() {
        guard levels.count > 1 else { return }
        for level in levels {
            let roomsHere = map.positions.filter { $0.cell.z == level }
            guard let top = roomsHere.compactMap({ frames[$0.room]?.minY }).min() else { continue }
            let title = level == 0 ? "Ground level" : (level > 0 ? "Level +\(level)" : "Level \(level)")
            let text = NSAttributedString(string: title.uppercased(), attributes: [
                .foregroundColor: Theme.foregroundFaint,
                .font: NSFont.systemFont(ofSize: 9, weight: .semibold),
                .kern: 0.8,
            ])
            text.draw(at: NSPoint(x: Self.margin, y: top - Self.bandHeader + 4))
        }
    }

    /// Draws every connection: a line within a level, a chevron pair across levels.
    private func drawConnections() {
        for connection in map.connections {
            guard connection.rooms.count == 2,
                  let a = frames[connection.rooms[0]],
                  let b = frames[connection.rooms[1]] else { continue }
            let levelA = level(of: connection.rooms[0])
            let levelB = level(of: connection.rooms[1])
            if levelA != levelB {
                drawLevelChange(from: a, to: b, rising: (levelB ?? 0) > (levelA ?? 0))
                continue
            }
            let path = NSBezierPath()
            path.move(to: NSPoint(x: a.midX, y: a.midY))
            path.line(to: NSPoint(x: b.midX, y: b.midY))
            path.lineWidth = 1.5
            (connection.via == nil ? Theme.worldConnection : Theme.worldDoor).setStroke()
            if connection.via != nil { path.setLineDash([4, 3], count: 2, phase: 0) }
            path.stroke()
        }
    }

    /// Marks a connection that leaves the band, on both of its rooms.
    /// - Parameters:
    ///   - a: the first room's frame
    ///   - b: the second room's frame
    ///   - rising: whether the second room sits on a higher level than the first
    private func drawLevelChange(from a: NSRect, to b: NSRect, rising: Bool) {
        let attributes: [NSAttributedString.Key: Any] = [
            .foregroundColor: Theme.worldConnection,
            .font: NSFont.systemFont(ofSize: 10, weight: .bold),
        ]
        NSAttributedString(string: rising ? "\u{25B2}" : "\u{25BC}", attributes: attributes)
            .draw(at: NSPoint(x: a.maxX - 12, y: a.minY + 2))
        NSAttributedString(string: rising ? "\u{25BC}" : "\u{25B2}", attributes: attributes)
            .draw(at: NSPoint(x: b.maxX - 12, y: b.minY + 2))
    }

    /// Draws every placed room.
    private func drawRooms() {
        let paragraph = NSMutableParagraphStyle()
        paragraph.alignment = .center
        paragraph.lineBreakMode = .byTruncatingTail

        for placed in map.positions {
            guard let box = frames[placed.room] else { continue }
            let isStart = placed.room == map.start
            let isUnreached = unreached.contains(placed.room)
            let isDisplaced = displaced.contains(placed.room)

            let shape = NSBezierPath(roundedRect: box.insetBy(dx: 0.5, dy: 0.5), xRadius: 5, yRadius: 5)
            (isUnreached ? Theme.worldRoomFillUnreached : Theme.worldRoomFill).setFill()
            shape.fill()
            if isDisplaced { shape.setLineDash([4, 3], count: 2, phase: 0) }
            shape.lineWidth = isStart ? 2 : 1
            (isStart ? Theme.accent : (isDisplaced ? Theme.worldDisplaced : Theme.border)).setStroke()
            shape.stroke()

            let color = isUnreached ? Theme.foregroundFaint : Theme.foreground
            let label = NSAttributedString(string: placed.room, attributes: [
                .foregroundColor: color,
                .font: NSFont.systemFont(ofSize: 11),
                .paragraphStyle: paragraph,
            ])
            label.draw(in: NSRect(x: box.minX + 4, y: box.midY - 8,
                                  width: box.width - 8, height: 16))
        }
    }

    /// The level a room sits on, or nil when it was never placed.
    /// - Parameter room: the room id
    /// - Returns: its z level
    private func level(of room: String) -> Int? {
        map.positions.first { $0.room == room }?.cell.z
    }
}

File: tools/ide/PaneHost/Shell/WorldMapView.cs
// WorldMapView.swift, drawn with a DrawingContext.
//
// Layout metrics from the shipping control (as recorded by the OpenSilver
// spike's Phase 4): 108×34 boxes, 24×22 gaps, 16 margin, 20 px band header,
// 18 px band gap, levels highest-first sharing one x origin. Doors dashed,
// start room stroked 2 px accent, displaced rooms dashed mauve, rooms play
// never reaches faded.
//
// Fed by the REAL analyzer: `sharpee compose` → `sharpee world-index`, the same
// JSON the IDE's World tab reads. Nothing here is synthetic.
//
// Public interface: WorldMapView (Load, RoomCount, ConnectionCount, DisplacedCount).
// Owner context: tools/ide — the Avalonia desktop head. Ported from the O6
// evaluation spike 2026-09-16; the evidence for its shape is that evaluation's record.

using System.Text.Json;
using Avalonia;
using Avalonia.Media;
using PaneHost.Theme;

namespace PaneHost.Shell;

/// <summary>One placed room: its compass cell and the flags the map colours by.</summary>
internal sealed record MapRoom(string Id, int X, int Y, int Z, bool IsStart, bool Displaced, bool Reached);

/// <summary>One passage between two rooms; <paramref name="Door"/> non-null means a door sits in it.</summary>
internal sealed record MapLink(string From, string To, string? Door);

/// <summary>The World tab's map, drawn from world-index output.</summary>
public sealed class WorldMapView : DrawnSurface
{
    private const double BoxWidth = 108, BoxHeight = 34;
    private const double GapX = 24, GapY = 22;
    private const double Pad = 16;
    private const double BandHeader = 20, BandGap = 18;

    private readonly List<MapRoom> _rooms = new();
    private readonly List<MapLink> _links = new();
    private readonly Dictionary<string, Rect> _boxes = new();
    private Size _content = new(400, 200);

    public int RoomCount => _rooms.Count;
    public int ConnectionCount => _links.Count;
    public int DisplacedCount => _rooms.Count(r => r.Displaced);
    public int LevelCount => _rooms.Select(r => r.Z).Distinct().Count();
    public int DoorCount => _links.Count(l => l.Door is not null);

    /// <summary>Reads a world-index document and lays the map out. Repaints.</summary>
    /// <exception cref="JsonException">The file is not world-index JSON.</exception>
    public void Load(string worldIndexPath)
    {
        using var json = JsonDocument.Parse(File.ReadAllText(worldIndexPath));
        var map = json.RootElement.GetProperty("map");
        var start = map.GetProperty("start").GetString();

        var displaced = map.GetProperty("collisions").EnumerateArray()
            .Select(c => c.GetProperty("room").GetString()!).ToHashSet();

        // reach.rooms is {total, reachable[], unreached[]} — a summary object, not a list.
        var reached = json.RootElement.TryGetProperty("reach", out var reach)
            && reach.TryGetProperty("rooms", out var reachRooms)
            && reachRooms.TryGetProperty("reachable", out var reachable)
            ? reachable.EnumerateArray().Select(r => r.GetString()!).ToHashSet()
            : new HashSet<string>();

        _rooms.Clear();
        foreach (var position in map.GetProperty("positions").EnumerateArray())
        {
            var id = position.GetProperty("room").GetString()!;
            var cell = position.GetProperty("cell");
            _rooms.Add(new MapRoom(
                id,
                cell.GetProperty("x").GetInt32(),
                cell.GetProperty("y").GetInt32(),
                cell.GetProperty("z").GetInt32(),
                id == start,
                displaced.Contains(id),
                reached.Count == 0 || reached.Contains(id)));
        }

        _links.Clear();
        foreach (var connection in map.GetProperty("connections").EnumerateArray())
        {
            var pair = connection.GetProperty("rooms").EnumerateArray().Select(r => r.GetString()!).ToArray();
            var via = connection.GetProperty("via");
            _links.Add(new MapLink(pair[0], pair[1], via.ValueKind == JsonValueKind.String ? via.GetString() : null));
        }

        LayoutBoxes();
        InvalidateVisual();
    }

    /// <summary>
    /// The Swift view's `layoutBoxes`, transliterated: one band per level, highest
    /// first, every band sharing one x origin so a room keeps its column across levels.
    /// </summary>
    private void LayoutBoxes()
    {
        _boxes.Clear();
        if (_rooms.Count == 0) { _content = new Size(400, 200); return; }

        var minX = _rooms.Min(r => r.X);
        var levels = _rooms.Select(r => r.Z).Distinct().OrderByDescending(z => z).ToList();
        var maxWidth = 0.0;
        var y = Pad;

        foreach (var level in levels)
        {
            y += BandHeader;
            var band = _rooms.Where(r => r.Z == level).ToList();
            var minY = band.Min(r => r.Y);
            var rows = band.Max(r => r.Y) - minY + 1;

            foreach (var room in band)
            {
                var left = Pad + (room.X - minX) * (BoxWidth + GapX);
                // Compass north is up, so a higher y cell sits nearer the top of its band.
                var top = y + (band.Max(r => r.Y) - room.Y) * (BoxHeight + GapY);
                _boxes[room.Id] = new Rect(left, top, BoxWidth, BoxHeight);
                maxWidth = Math.Max(maxWidth, left + BoxWidth + Pad);
            }

            y += rows * (BoxHeight + GapY) + BandGap;
        }

        _content = new Size(maxWidth, y + Pad);
    }

    protected override Size MeasureOverride(Size availableSize) => _content;

    protected override void Draw(DrawingContext context)
    {
        context.FillRectangle(ThemeTokens.EditorBackground, new Rect(Bounds.Size));
        if (_rooms.Count == 0) return;

        // Band headers first, so boxes and links draw over them.
        var levels = _rooms.Select(r => r.Z).Distinct().OrderByDescending(z => z).ToList();
        foreach (var level in levels)
        {
            var band = _rooms.Where(r => r.Z == level).ToList();
            var top = band.Min(r => _boxes[r.Id].Y) - BandHeader;
            var label = Text(level == 0 ? "ground" : level > 0 ? $"level +{level}" : $"level {level}",
                10.5, ThemeTokens.ForegroundFaint);
            context.DrawText(label, new Point(Pad, top));
        }

        var connection = new Pen(ThemeTokens.WorldConnection, 1.5);
        var door = new Pen(ThemeTokens.WorldDoor, 1.5, new DashStyle(new double[] { 4, 3 }, 0));

        foreach (var link in _links)
        {
            if (!_boxes.TryGetValue(link.From, out var a) || !_boxes.TryGetValue(link.To, out var b)) continue;
            context.DrawLine(link.Door is null ? connection : door, a.Center, b.Center);
        }

        foreach (var room in _rooms)
        {
            var box = _boxes[room.Id];
            var fill = room.Reached ? ThemeTokens.WorldRoomFill : ThemeTokens.WorldRoomFillUnreached;
            var pen = room switch
            {
                { IsStart: true } => new Pen(ThemeTokens.Accent, 2),
                { Displaced: true } => new Pen(ThemeTokens.WorldDisplaced, 1.5, new DashStyle(new double[] { 3, 2 }, 0)),
                { Reached: false } => new Pen(ThemeTokens.WorldUnreached, 1),
                _ => new Pen(ThemeTokens.Border, 1),
            };
            context.DrawRectangle(fill, pen, new RoundedRect(box, 5));

            var name = Text(room.Id, 10.5, room.Reached ? ThemeTokens.Foreground : ThemeTokens.WorldUnreached);
            var clipped = Math.Min(name.Width, BoxWidth - 10);
            using (context.PushClip(box.Deflate(5)))
                context.DrawText(name, new Point(box.X + (BoxWidth - clipped) / 2, box.Y + (BoxHeight - name.Height) / 2));
        }
    }
}
